File: client/src/castleCollision.ts
import { publicAssetPath } from './publicAssets';
import {
  CASTLE_COLLISION_BIN_PUBLIC_PATH,
  CASTLE_COLLISION_VERSION,
} from './castleCollisionMeta';

const HEADER_BYTES = 232;

export interface CastleCollisionAsset {
  readonly sourceNodeName: string;
  readonly sourceHash: string;
  readonly min: readonly [number, number, number];
  readonly max: readonly [number, number, number];
  readonly terrainScale: number;
  readonly terrainOffset: readonly [number, number, number];
  readonly sourceNodeMatrix: readonly number[];
  readonly vertices: Float32Array;
  readonly indices: Uint32Array;
  readonly grid: {
    readonly x: number;
    readonly y: number;
    readonly z: number;
    readonly offsets: Uint32Array;
    readonly triangleIds: Uint32Array;
  };
}

let collision: CastleCollisionAsset | null = null;
let loadPromise: Promise<void> | null = null;

function readAscii(view: DataView, start: number, length: number): string {
  let value = '';
  for (let index = 0; index < length; index += 1) {
    const char = view.getUint8(start + index);
    if (char === 0) break;
    value += String.fromCharCode(char);
  }
  return value;
}

function readHash(view: DataView, start: number): string {
  return Array.from({ length: 32 }, (_, index) => view.getUint8(start + index).toString(16).padStart(2, '0')).join('');
}

/** Decode CC01. The grid only chooses stable broad-phase candidates; all contacts use the original indexed triangles. */
export function loadCastleCollisionFromArrayBuffer(buffer: ArrayBuffer): void {
  if (buffer.byteLength < HEADER_BYTES) {
    throw new Error(`Castle collision buffer too short: ${buffer.byteLength} bytes`);
  }
  const view = new DataView(buffer);
  if (readAscii(view, 0, 4) !== 'CC01') {
    throw new Error(`Bad castle collision magic: ${readAscii(view, 0, 4)}`);
  }
  const version = view.getUint32(4, true);
  if (version !== CASTLE_COLLISION_VERSION) {
    throw new Error(`Castle collision version ${version} != expected ${CASTLE_COLLISION_VERSION}`);
  }
  const headerBytes = view.getUint32(8, true);
  if (headerBytes !== HEADER_BYTES) {
    throw new Error(`Castle collision header ${headerBytes} != expected ${HEADER_BYTES}`);
  }
  const vertexCount = view.getUint32(12, true);
  const triangleCount = view.getUint32(16, true);
  const gridX = view.getUint32(20, true);
  const gridY = view.getUint32(24, true);
  const gridZ = view.getUint32(28, true);
  if (!vertexCount || !triangleCount || !gridX || !gridY || !gridZ) {
    throw new Error('Castle collision contains an empty mesh or grid');
  }

  const min: [number, number, number] = [view.getFloat32(32, true), view.getFloat32(36, true), view.getFloat32(40, true)];
  const max: [number, number, number] = [view.getFloat32(44, true), view.getFloat32(48, true), view.getFloat32(52, true)];
  const terrainScale = view.getFloat32(56, true);
  const terrainOffset: [number, number, number] = [view.getFloat32(60, true), view.getFloat32(64, true), view.getFloat32(68, true)];
  const sourceNodeMatrix = Array.from({ length: 16 }, (_, index) => view.getFloat32(72 + index * 4, true));
  const sourceHash = readHash(view, 136);
  const sourceNodeName = readAscii(view, 168, 64);
  const cellCount = gridX * gridY * gridZ;
  if (!Number.isSafeInteger(cellCount)) throw new Error('Castle collision grid dimensions overflow');
  if (!min.every(Number.isFinite) || !max.every(Number.isFinite) || min.some((value, axis) => value >= max[axis])) {
    throw new Error('Castle collision bounds are not finite and ordered');
  }
  const verticesOffset = headerBytes;
  const indicesOffset = verticesOffset + vertexCount * 3 * 4;
  const offsetsOffset = indicesOffset + triangleCount * 3 * 4;
  const triangleIdsOffset = offsetsOffset + (cellCount + 1) * 4;
  if (buffer.byteLength < triangleIdsOffset) {
    throw new Error('Castle collision buffer truncated before grid offsets');
  }
  const triangleIdCount = view.getUint32(offsetsOffset + cellCount * 4, true);
  const requiredBytes = triangleIdsOffset + triangleIdCount * 4;
  if (buffer.byteLength !== requiredBytes) {
    throw new Error(`Castle collision buffer truncated: need ${requiredBytes}, got ${buffer.byteLength}`);
  }

  const vertices = new Float32Array(vertexCount * 3);
  const indices = new Uint32Array(triangleCount * 3);
  const offsets = new Uint32Array(cellCount + 1);
  const triangleIds = new Uint32Array(triangleIdCount);
  for (let index = 0; index < vertices.length; index += 1) vertices[index] = view.getFloat32(verticesOffset + index * 4, true);
  for (let index = 0; index < indices.length; index += 1) indices[index] = view.getUint32(indicesOffset + index * 4, true);
  for (let index = 0; index < offsets.length; index += 1) offsets[index] = view.getUint32(offsetsOffset + index * 4, true);
  for (let index = 0; index < triangleIds.length; index += 1) triangleIds[index] = view.getUint32(triangleIdsOffset + index * 4, true);
  if (offsets[offsets.length - 1] !== triangleIds.length) {
    throw new Error('Castle collision grid offset count does not match triangle ids');
  }
  if (indices.some(index => index >= vertexCount)) throw new Error('Castle collision index outside vertex array');
  if (triangleIds.some(id => id >= triangleCount)) throw new Error('Castle collision grid references a missing triangle');
  if (offsets.some((offset, index) => index > 0 && offset < offsets[index - 1])) {
    throw new Error('Castle collision grid offsets are not monotonic');
  }

  collision = { sourceNodeName, sourceHash, min, max, terrainScale, terrainOffset, sourceNodeMatrix, vertices, indices, grid: { x: gridX, y: gridY, z: gridZ, offsets, triangleIds } };
}

export function initCastleCollision(url = publicAssetPath(CASTLE_COLLISION_BIN_PUBLIC_PATH)): Promise<void> {
  if (collision) return Promise.resolve();
  if (!loadPromise) {
    loadPromise = fetch(url)
      .then(response => {
        if (!response.ok) throw new Error(`Failed to fetch castle collision: ${response.status} ${url}`);
        return response.arrayBuffer();
      })
      .then(loadCastleCollisionFromArrayBuffer)
      .catch(error => {
        loadPromise = null;
        throw error;
      });
  }
  return loadPromise;
}

export function castleCollisionAsset(): CastleCollisionAsset {
  if (!collision) throw new Error('Castle collision not loaded — call await initCastleCollision() before querying');
  return collision;
}

export function isCastleCollisionReady(): boolean {
  return collision !== null;
}

/** Returns ascending triangle ids for a world-space AABB, with duplicate grid hits removed. */
export function castleTriangleCandidates(min: readonly number[], max: readonly number[]): readonly number[] {
  const asset = castleCollisionAsset();
  if (
    max[0] < asset.min[0] || min[0] > asset.max[0]
    || max[1] < asset.min[1] || min[1] > asset.max[1]
    || max[2] < asset.min[2] || min[2] > asset.max[2]
  ) {
    return [];
  }
  const { grid } = asset;
  const dims = [grid.x, grid.y, grid.z];
  const start = [0, 1, 2].map(axis => gridCoordinate(min[axis], asset.min[axis], asset.max[axis], dims[axis]));
  const end = [0, 1, 2].map(axis => gridCoordinate(max[axis], asset.min[axis], asset.max[axis], dims[axis]));
  const candidates = new Set<number>();
  for (let z = start[2]; z <= end[2]; z += 1) {
    for (let y = start[1]; y <= end[1]; y += 1) {
      for (let x = start[0]; x <= end[0]; x += 1) {
        const cell = (z * grid.y + y) * grid.x + x;
        for (let index = grid.offsets[cell]; index < grid.offsets[cell + 1]; index += 1) candidates.add(grid.triangleIds[index]);
      }
    }
  }
  return [...candidates].sort((left, right) => left - right);
}

function gridCoordinate(value: number, min: number, max: number, dimension: number): number {
  if (value <= min) return 0;
  if (value >= max) return dimension - 1;
  return Math.min(dimension - 1, Math.max(0, Math.floor(((value - min) / (max - min)) * dimension)));
}

File: server/spacetimedb/src/castle_collision.rs
use std::sync::OnceLock;

const MAGIC: &[u8; 4] = b"CC01";
const VERSION: u32 = 1;
const HEADER_BYTES: usize = 232;

/// Canonical transformed triangles for `Castle Collision.002`, baked from the active map GLB.
/// The grid selects deterministic broad-phase candidates only; callers must run narrow phase
/// against `vertices` and `indices` for every returned id.
static CASTLE_COLLISION_BIN: &[u8] = include_bytes!("castle_collision.bin");

pub struct CastleCollisionData {
    pub min: [f32; 3],
    pub max: [f32; 3],
    pub vertices: Vec<f32>,
    pub indices: Vec<u32>,
    grid_x: usize,
    grid_y: usize,
    grid_z: usize,
    offsets: Vec<u32>,
    triangle_ids: Vec<u32>,
}

static CASTLE_COLLISION: OnceLock<CastleCollisionData> = OnceLock::new();

pub fn castle_collision() -> &'static CastleCollisionData {
    CASTLE_COLLISION.get_or_init(|| {
        parse_castle_collision_bin(CASTLE_COLLISION_BIN).expect("valid embedded castle_collision.bin")
    })
}

fn parse_castle_collision_bin(bytes: &[u8]) -> Result<CastleCollisionData, String> {
    if bytes.len() < HEADER_BYTES {
        return Err("castle_collision.bin too short".into());
    }
    if &bytes[0..4] != MAGIC {
        return Err("bad castle collision magic".into());
    }
    if read_u32(bytes, 4)? != VERSION {
        return Err("unsupported castle collision version".into());
    }
    if read_u32(bytes, 8)? as usize != HEADER_BYTES {
        return Err("unexpected castle collision header size".into());
    }
    let vertex_count = read_u32(bytes, 12)? as usize;
    let triangle_count = read_u32(bytes, 16)? as usize;
    let grid_x = read_u32(bytes, 20)? as usize;
    let grid_y = read_u32(bytes, 24)? as usize;
    let grid_z = read_u32(bytes, 28)? as usize;
    if vertex_count == 0 || triangle_count == 0 || grid_x == 0 || grid_y == 0 || grid_z == 0 {
        return Err("castle collision has an empty mesh or grid".into());
    }
    let min = [read_f32(bytes, 32)?, read_f32(bytes, 36)?, read_f32(bytes, 40)?];
    let max = [read_f32(bytes, 44)?, read_f32(bytes, 48)?, read_f32(bytes, 52)?];
    let cell_count = grid_x
        .checked_mul(grid_y)
        .and_then(|value| value.checked_mul(grid_z))
        .ok_or("castle collision grid dimensions overflow")?;
    let vertices_offset = HEADER_BYTES;
    let indices_offset = vertices_offset
        .checked_add(vertex_count.checked_mul(12).ok_or("castle collision vertices overflow")?)
        .ok_or("castle collision indices offset overflow")?;
    let offsets_offset = indices_offset
        .checked_add(triangle_count.checked_mul(12).ok_or("castle collision indices overflow")?)
        .ok_or("castle collision offsets overflow")?;
    let triangle_ids_offset = offsets_offset
        .checked_add((cell_count + 1).checked_mul(4).ok_or("castle collision offsets overflow")?)
        .ok_or("castle collision triangle ids offset overflow")?;
    if bytes.len() < triangle_ids_offset {
        return Err("castle_collision.bin truncated before grid offsets".into());
    }
    let triangle_id_count = read_u32(bytes, offsets_offset + cell_count * 4)? as usize;
    let required_bytes = triangle_ids_offset
        .checked_add(triangle_id_count.checked_mul(4).ok_or("castle collision triangle ids overflow")?)
        .ok_or("castle collision binary length overflow")?;
    if bytes.len() < required_bytes {
        return Err("castle_collision.bin truncated".into());
    }

    let vertices = read_u32_values(bytes, vertices_offset, vertex_count * 3)?
        .into_iter()
        .map(f32::from_bits)
        .collect();
    let indices = read_u32_values(bytes, indices_offset, triangle_count * 3)?;
    let offsets = read_u32_values(bytes, offsets_offset, cell_count + 1)?;
    let triangle_ids = read_u32_values(bytes, triangle_ids_offset, triangle_id_count)?;
    if offsets.last().copied() != Some(triangle_ids.len() as u32) {
        return Err("castle collision final grid offset does not match triangle ids".into());
    }
    if offsets.windows(2).any(|pair| pair[0] > pair[1]) {
        return Err("castle collision grid offsets are not monotonic".into());
    }
    if indices.iter().any(|&index| index as usize >= vertex_count) {
        return Err("castle collision index outside vertex array".into());
    }
    if triangle_ids.iter().any(|&id| id as usize >= triangle_count) {
        return Err("castle collision grid references a missing triangle".into());
    }

    Ok(CastleCollisionData { min, max, vertices, indices, grid_x, grid_y, grid_z, offsets, triangle_ids })
}

impl CastleCollisionData {
    /// Candidate triangle ids are ascending after deduplication, ensuring a stable narrow-phase order.
    pub fn triangle_candidates(&self, min: [f32; 3], max: [f32; 3]) -> Vec<u32> {
        if max[0] < self.min[0] || min[0] > self.max[0]
            || max[1] < self.min[1] || min[1] > self.max[1]
            || max[2] < self.min[2] || min[2] > self.max[2]
        {
            return Vec::new();
        }
        let start = [
            self.grid_coordinate(min[0], self.min[0], self.max[0], self.grid_x),
            self.grid_coordinate(min[1], self.min[1], self.max[1], self.grid_y),
            self.grid_coordinate(min[2], self.min[2], self.max[2], self.grid_z),
        ];
        let end = [
            self.grid_coordinate(max[0], self.min[0], self.max[0], self.grid_x),
            self.grid_coordinate(max[1], self.min[1], self.max[1], self.grid_y),
            self.grid_coordinate(max[2], self.min[2], self.max[2], self.grid_z),
        ];
        let mut candidates = Vec::new();
        for z in start[2]..=end[2] {
            for y in start[1]..=end[1] {
                for x in start[0]..=end[0] {
                    let cell = (z * self.grid_y + y) * self.grid_x + x;
                    candidates.extend_from_slice(&self.triangle_ids[self.offsets[cell] as usize..self.offsets[cell + 1] as usize]);
                }
            }
        }
        candidates.sort_unstable();
        candidates.dedup();
        candidates
    }

    fn grid_coordinate(&self, value: f32, min: f32, max: f32, dimension: usize) -> usize {
        if value <= min { return 0; }
        if value >= max { return dimension - 1; }
        (((value - min) / (max - min)) * dimension as f32).floor() as usize
    }
}

fn read_u32(bytes: &[u8], offset: usize) -> Result<u32, String> {
    let end = offset.checked_add(4).ok_or("castle collision offset overflow")?;
    let slice: [u8; 4] = bytes.get(offset..end).ok_or("castle collision read out of bounds")?.try_into().map_err(|_| "invalid castle collision u32")?;
    Ok(u32::from_le_bytes(slice))
}

fn read_f32(bytes: &[u8], offset: usize) -> Result<f32, String> {
    Ok(f32::from_bits(read_u32(bytes, offset)?))
}

fn read_u32_values(bytes: &[u8], offset: usize, count: usize) -> Result<Vec<u32>, String> {
    (0..count).map(|index| read_u32(bytes, offset + index * 4)).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn embedded_asset_has_exact_triangle_data_and_a_nonempty_grid() {
        let data = castle_collision();
        assert!(data.vertices.len() > 100_000);
        assert!(data.indices.len() > 100_000);
        assert!(data.max[0] > data.min[0]);
        assert!(data.max[1] > data.min[1]);
        assert!(data.max[2] > data.min[2]);
    }

    #[test]
    fn broad_phase_returns_stably_sorted_ids_inside_castle_bounds() {
        let data = castle_collision();
        let candidates = data.triangle_candidates(data.min, data.max);
        assert!(!candidates.is_empty());
        assert!(candidates.windows(2).all(|pair| pair[0] < pair[1]));
    }
}
